util/webhook/validator.go

```diff
@@ -136,3 +144,43 @@ func validationResponseFromStatus(allowed bool, status metav1.Status) admission.
 		},
 	}
 }
+
+// ValidateImmutable validates equality across two values,
+// and returns a meaningful error to indicate a changed value, a newly set value, or a newly unset value.
+func ValidateImmutable(path *field.Path, oldVal, newVal any) *field.Error {
```

Comment:
This is definitely much less verbose and repetitive than having separate funcs for every type, though I can see where this introduces some foot-guns like using this for a not-explicitly-listed type and reasoning about how fields of something like `type VMIdentity string` behave.

Overall I could be convinced to start looking into generics for this function since I think that would address most of those concerns.

Reply:
I'd love to see the go generic equivalent for this!

Reply:
For reference, I took a pass at what that might look like and came up with this:
```go
func ValidateImmutable[T any](path *field.Path, oldVal, newVal T) *field.Error {
	var unsetVal T
	if !reflect.DeepEqual(oldVal, unsetVal) {
		// Prevent modification if it was already set to some value
		if reflect.DeepEqual(newVal, unsetVal) {
			// unsetting the field is not allowed
			return field.Invalid(path, newVal, unsetMessage)
		}
		if !reflect.DeepEqual(oldVal, newVal) {
			// changing the field is not allowed
			return field.Invalid(path, newVal, immutableMessage)
		}
	} else if !reflect.DeepEqual(newVal, unsetVal) {
		return field.Invalid(path, newVal, setMessage)
	}

	return nil
}
```

Reply:
Actually, [`reflect.IsZero()`](https://pkg.go.dev/reflect#Value.IsZero) might make this just as easy but without generics at all.

Reply:
IsZero ftw, thanks for the suggestion!